### ISSUE TITLE
normalizeNumbers panic

### ISSUE THREAD

**Issue by xakepp35** [itchyny/gojq#212]
It looks like `normalizeNumbers` function is a problem, when [unavoidably called from RunWithContext](https://github.com/itchyny/gojq/blob/main/compiler.go#L52)

Please, check out my reproduction example on [go.dev](https://go.dev/play/p/2LITP8PC7GM)

Is there any option to export it and let package user to decide wherever to call it, or not at all?

To rephrase - RunWithContext should be able to perform as a "pure" function - produce output, while never touching user input at all, because you cant know, if its safe to do any modification with it

Can RunWithContext behave as a "copy-on-write" executor, along whole execution path?

**Comment by itchyny** [comment]
Delaying the normalization could affect the package users adding custom internal functions with `gojq.WithFunction`.

**Comment by wader** [comment]
Possible to make it a `gojq.WithInputNormalize(<bool>)` compile option etc? this would also apply to `gojq.WithVariables` values i guess?

**Comment by itchyny** [comment]
At this moment, I have no good idea to solve this issue without increasing the number of parsing or breaking the custom internal function APIs. Also, I think in the most use case of this package (like gh), only one querying is executed against one input, so I can't imagine specific application that needs to reuse the same input value.

**Comment by reyoung** [comment]
I actually face the same issue last year when using `jq` as a script language in an internal system.

I just patch JQ, and manually copy `map` and `slice` instead.

However, it create a lot of memory and increase GC pressure.

https://github.com/reyoung/gojq/commit/19cb611aa3878c3faca43d83ba84cb1030659d6a

**Comment by itchyny** [comment]
I have pushed an experimental fix for this issue to `json-number` branch. You can install with `go install github.com/itchyny/gojq@json-number`.

**Comment by xakepp35** [comment]
@itchyny would json-number be merged into master? looks like it resolved our issue, using it in prod for 2 months, no problems were observed yet. moreover it allowed us to find and see the problems that we had with types on our side. can we merge and close this one then?

**Comment by itchyny** [comment]
This issue is blocked by https://github.com/go-yaml/yaml/issues/807.
